### DISCUSSION TITLE
c++ / python director and Eigen::VectorXd

### DISCUSSION THREAD

**Discussion by ArnaultT** [swig/swig#3080] [Q&A]
Hi everyone,

I have a virtual C++ class that I want to extend in Python.
This class contains a few computational functions that need to be overloaded.
For instance:

```c++
class myCppClass
{
public: 
    myCppClass(){}
    virtual Eigen::vectorXd compute_things();   
    void update_class()
   {
      // ... do things 
      toto = compute_things(); 
      // ... do other things
   }

protected: 
    Eigen::vectorXd _toto; 

};
```

The idea is that users implement their own `compute_things` method in Python. This method is called inside the `update_class` method.
To achieve this, I am using SWIG's director feature along with Eigen typemaps developed by [this repository](https://github.com/rdeits/swig-eigen-numpy/tree/master/swigmake/swig/python).

Here is what my .i file looks like:

```
%module(package="example", directors="1", allprotected="1", "threads"=1) My_Example

%{
#include <MyCppClass.h>
}

%import <typemaps.i>
%include <std_vector.i>
%include <eigen.i>
%import "numpy.i"

%feature("director") myCppClass;

%include <MyCppClass.h>
```

Then in python : 
```python
class myPythonClass(myCppClass):

   def __init__(self): 
       super().__init__()

   def compute_things(self): 
       return np.array([2.0, 1.0])


myclass = myPythonClass()
myclass.upadate_class()
```

When I run this code, I get the following error:
"TypeError: SWIG director type mismatch in output value of type 'Eigen::VectorXd'"

However, if I call the `compute_things` method directly, it works without any issue.

Some additional context:
- In my complete code, the instance of `myPythonClass` is __disown__()-ed, but I cannot share the full implementation here.
- If I replace `Eigen::VectorXd` with more "standard" types such as `std::vector<double>`, `int`, or `double`, everything works fine.
- I tried adding debug print statements in all the Eigen typemaps. It seems that the TypeError occurs before the typemaps are triggered, which leaves me unsure of where the issue lies.

Is this a limitation of the SWIG-Eigen integration, or am I doing something wrong? Any help would be greatly appreciated!

Thank you very much!

**Comment by wsfulton** [comment]
I don't know how vectorXd is being handled/marshalled to C++. However, it looks like you expect a numpy.array type to convert to it automatically. I suspect you need a directorout typemap in order to implement this logic when using directors and the return type in the director method:
```c++
    virtual Eigen::vectorXd compute_things();  
```
I suspect the director typemaps are not provided.
